src/macaron/slsa_analyzer/provenance/loader.py

```diff
@@ -1,48 +1,71 @@
+    except (configparser.Error, UnicodeDecodeError):
+        return None
 
-    Inside a provenance file is a DSSE envelope containing a base64-encoded
-    provenance JSON payload. See: https://github.com/secure-systems-lab/dsse.
-    If the file is gzipped, it will be transparently decompressed.
 
-    Parameters
-    ----------
-    filepath : str
-        Path to the provenance file.
+def _download_url_file_content(url: str, url_link_hostname_allowlist: list[str]) -> bytes:
+    hostname = urlparse(url).hostname
+    if hostname is None or hostname == "":
+        raise LoadIntotoAttestationError("Cannot resolve URL link file: invalid URL")
+    if hostname not in url_link_hostname_allowlist:
+        raise LoadIntotoAttestationError(
+            "Cannot resolve URL link file: target hostname '" + hostname + "' is not in allowed hostnames."
+        )
 
-    Returns
-    -------
-    dict[str, JsonType]
-        The provenance JSON payload.
+    # TODO download size limit?
+    timeout = defaults.getint("downloads", "timeout", fallback=120)
+    response = send_get_http_raw(url=url, headers=None, timeout=timeout)
+    if response is None:
+        raise LoadIntotoAttestationError("Cannot resolve URL link file: Failed to download file")
+    if response.status_code != 200:
+        raise LoadIntotoAttestationError(
+            "Cannot resolve URL link file: Failed to download file, error " + str(response.status_code)
+        )
+    return response.content
+
+
+def _load_provenance_file_content(
+    file_content: bytes, url_link_hostname_allowlist: list[str], url_link_depth_limit: int = 5
+) -> dict[str, JsonType]:
+    url_link_depth = 0
+    while url_link_depth <= url_link_depth_limit:
+        url = _try_read_url_link_file(file_content)
+        if url is None:
+            break
+        if url_link_depth == url_link_depth_limit:
+            raise LoadIntotoAttestationError("Cannot resolve URL link file: depth limit exceeded")
+        file_content = _download_url_file_content(url, url_link_hostname_allowlist)
+        url_link_depth = url_link_depth + 1
 
-    Raises
-    ------
-    LoadIntotoAttestationError
-        If there is an error loading the provenance JSON payload.
-    """
     try:
         try:
-            with gzip.open(filepath, mode="rt", encoding="utf-8") as file:
-                provenance = json.load(file)
-        except (gzip.BadGzipFile, EOFError, zlib.error):
-            with open(filepath, encoding="utf-8") as file:
-                provenance = json.load(file)
-    except (OSError, json.JSONDecodeError, TypeError) as error:
+            decompressed_file_content = gzip.decompress(file_content)
+            provenance = json.loads(decompressed_file_content.decode())
+        except (gzip.BadGzipFile, EOFError, zlib.error, configparser.NoOptionError):
+            provenance = json.loads(file_content.decode())
+    except (json.JSONDecodeError, TypeError) as error:
```

Comment:
Should we catch `UnicodeDecodeError` here similar to https://github.com/oracle/macaron/blob/3e5e27d363ec76afafc6dd206c8d09aff71551bc/src/macaron/slsa_analyzer/provenance/loader.py#L24 because we are decoding `decompressed_file_content` which is of type `bytes`?

Reply:
Yes we should. Done.